src/main/java/edu/ucla/library/iiif/fester/handlers/PostThumbnailsHandler.java

```diff
@@ -0,0 +1,292 @@
+                }
+                promise.complete();
+            } else {
+                promise.fail(asyncResult.result().statusMessage());
+            }
+        });
+
+        return promise.future();
+
+    }
+
+    /**
+     * Retireve the base thumbnail URL from IIIF V2 presentation manifests.
+     *
+     * @param aColumnIndex The column in the CSV where thumbnail will be added
+     * @param aRowIndex The CSV row being updated
+     * @param aManifest A IIIF work manifest
+     * @param aCsvList A CSV file parsed as a list of string arrays
+     */
+    private void addV2Thumb(final int aColumnIndex, final int aRowIndex,
+                            final JsonObject aManifest, final List<String[]> aCsvList) {
+        final JsonArray canvases = aManifest.getJsonArray(SEQUENCES).getJsonObject(0).getJsonArray(CANVASES);
+        final int canvasIndex = chooseThumbIndex(canvases.size());
+        final String thumbURL = canvases.getJsonObject(canvasIndex).getJsonArray(IMAGE_CONTENT)
+                                .getJsonObject(0).getJsonObject(RESOURCE).getJsonObject(SERVICE)
+                                .getString(Constants.ID_V2);
+        ThumbnailUtils.addThumbnailURL(aColumnIndex, aRowIndex, thumbURL, aCsvList);
+    }
+
+    /**
+     * Retireve the base thumbnail URL from IIIF V3 presentation manifests.
+     *
+     * @param aColumnIndex The column in the CSV where thumbnail will be added
+     * @param aRowIndex The CSV row being updated
+     * @param aManifest A IIIF work manifest
+     * @param aCsvList A CSV file parsed as a list of string arrays
+     */
+    private void addV3Thumb(final int aColumnIndex, final int aRowIndex,
+                            final JsonObject aManifest, final List<String[]> aCsvList) {
+
+        final JsonArray canvases = aManifest.getJsonArray(ITEMS).getJsonObject(0).getJsonArray(ITEMS);
+        final int canvasIndex = chooseThumbIndex(canvases.size());
+        final int imageIndex = chooseThumbIndex(canvases.getJsonObject(canvasIndex).getJsonArray(ITEMS).size());
+        final String thumbURL = canvases.getJsonObject(canvasIndex).getJsonArray(ITEMS).getJsonObject(imageIndex)
+                               .getJsonObject(BODY).getJsonArray(SERVICE).getJsonObject(0).getString(Constants.ID_V3);
+
+        ThumbnailUtils.addThumbnailURL(aColumnIndex, aRowIndex, thumbURL, aCsvList);
+    }
+
+    /**
+     * Select the index for the thumbnail from a list of images.
+     *
+     * @param aCount The number of images
+     */
+    private int chooseThumbIndex(final int aCount) {
+        if (aCount <= 3) {
+            return 0;
+        } else {
+            return ThumbnailUtils.pickThumbnailIndex(aCount - 1);
+        }
+    }
+
+    /**
+     * Return the processed CSV to user.
+     *
+     * @param aFileName Name of CSV file
+     * @param aFilePath Path to CSV file
+     * @param aCsvList CSV file paresed as list of string arrays
```

Comment:
*parsed

Reply:
Dang it, Jim, I'm a coder, not a speller! /s
Fixed.